### PR TITLE
RFC? Refactor stat

### PR DESCRIPTION
**Pull request by bramtayl**
When doing #25188, I noticed there was a lot of unused filesystem functions, and mostly stat ones. Though there didn't seem to be much interest in #25939, and its probably too late in the release cycle, I thought I'd put up a proposal anyway. This is totally incomplete, with only direct changes to the code. I thought I'd see what people think before going through the effort of cleaning up.

### CODE REVIEW

**Comment by bramtayl** [comment]
Summary:

- `lstat` refactored into `stat`.
- `filemode`, `filesize`, `mtime`, and `ctime` deprecated in favor of field accesses of `stat`.
- `ispath`, `isfifo`, `ischardev`, `isdir`, `isblockdev`, `isfile`, `islink`, `issocket`, and `issetuid` refactored to `filetype`.
- `issetuid`, `issetgid`, and `issticky` refactored to `fileflags`.
- `uperm`, `operm`, and `gperm` refactored to `permissions`.

File: base/stat.jl [hunk 1]
```diff
@@ -104,182 +108,97 @@ The fields of the structure are:
...
+translate_permission(bit) =
+    if bit == 1
+        :execute
+    elseif bit == 2
+        :write
+    elseif bit == 4
+        :read
+    else
+        error("Unknown permissions type")
+    end
```
Comment (JeffBezanson) [inline base/stat.jl]: This doesn't handle multiple permission bits being set.

Comment (bramtayl) [reply base/stat.jl]: Oh, ok, I must have misunderstood the code. How would I fix it?

Comment (bramtayl) [reply base/stat.jl]: I thought I would be ok with `permissions` returning a named tuple for all 3.

**Comment by StefanKarpinski** [comment]
This is a pretty nice redesign. “Neither” is more commonly called “other”.

File: base/stat.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/stat.jl]: That part is ok, but each of user/group/other might have any combination of read/write/execute.

File: base/stat.jl [hunk 2]
```diff
@@ -83,9 +83,13 @@ end
 stat(fd::Integer)           = stat(RawFD(fd))
 
 """
-    stat(file)
+    stat(file; link = false)
```
Comment (ararslan) [inline base/stat.jl]: no spaces around = for keyword arguments

File: base/stat.jl [hunk 3]
```diff
@@ -104,182 +108,97 @@ The fields of the structure are:
...
-issetuid(st::StatStruct) = (filemode(st) & 0o4000) > 0
-
-"""
-    issetgid(path) -> Bool
+function stat(path..., link = false)
+    full_path = joinpath(path...)
+    if link
+        stat(full_path)
+    else
+         lstat(full_path)
```
Comment (ararslan) [inline base/stat.jl]: One too many spaces

**Comment by fredrikekre** [comment]
I really *really* dislike the change from `isdir(f)` to `filetype(f) == :dir`. And `ispath(f)` to `filetype(f) != :invalid` is even worse.

**Comment by ararslan** [comment]
Most of this doesn't really seem like an improvement to me, tbh.

**Comment by KristofferC** [comment]
This PR is taking on too many things at once imo. The getfield on the `StatStruct` seems OK but some of the other things (like the `filetype` seems extremely verbose and brittle (what if you misspell the symbol slightly). It should use enums if anything.

**Comment by bramtayl** [comment]
I'd be ok with keeping a few convenience functions (`ispath` is definitely a keeper). I'm not sure I see the advantage of filetype returning enums rather than symbols but I could change that too.

**Comment by StefanKarpinski** [comment]
Honestly, I think it's too late for this. We're well past feature freeze so any API that's not deeply broken is not up for change at this point. Maybe in Julia 2.0 or we could have an _additional_ better API here and keep the old one, but that would likely go into a later 1.x release, not 1.0.

**Pull request closed without merging**